## Загрузка и подготовка данных

Импортируем все необходимые для работы библиотеки и инструменты

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

In [2]:
geo_data_0 = pd.read_csv('/datasets/geo_data_0.csv')
geo_data_1 = pd.read_csv('/datasets/geo_data_1.csv')
geo_data_2 = pd.read_csv('/datasets/geo_data_2.csv')

Рассмотрим первые 5 строк датасетов и информацию о них.

In [3]:
def first_view(data, number):
    return (
        print(""),
        print("\033[1m" + "Первые 5 строк датасета geo_data_", number, "." + "\033[0m"),
        display(data.head()),
        print("\033[1m" + "Информация о датасете geo_data_", number, "." + "\033[0m"),
        print(""),
        print(data.info()),
        print("\033[1m" + "Пропуски в датасете geo_data_", number, "." + "\033[0m"),
        print(""),
        print(data.isna().sum()),
        print("\033[1m" + "Статистика для датасета geo_data_", number, "." + "\033[0m"),
        print(""),
        print(data.describe()),
        print(
            "\033[1m" + "Сумма уникальных значений признаков для датасета geo_data_",
            number,
            "." + "\033[0m",
        ),
        print(""),
        print(data.nunique()),
    )

In [4]:
first_view(geo_data_0, 0)
first_view(geo_data_1, 1)
first_view(geo_data_2, 2)


Первые 5 строк датасета geo_data_ 0 .


id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

Информация о датасете geo_data_ 0 .

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None
Пропуски в датасете geo_data_ 0 .

id         0
f0         0
f1         0
f2         0
product    0
dtype: int64
Статистика для датасета geo_data_ 0 .

                  f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.500419       0.250143       2.502647      92.500000
std         0.871832       0.504433       3.248248      44.288691
min        -1.408605      -0.848218     -12.088328       0.000000
25%        -0.072580      -0.200881       

id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305

Информация о датасете geo_data_ 1 .

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None
Пропуски в датасете geo_data_ 1 .

id         0
f0         0
f1         0
f2         0
product    0
dtype: int64
Статистика для датасета geo_data_ 1 .

                  f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        1.141296      -4.796579       2.494541      68.825000
std         8.965932       5.119872       1.703572      45.944423
min       -31.609576     -26.358598      -0.018144       0.000000
25%        -6.298551      -8.267985       

id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746

Информация о датасете geo_data_ 2 .

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None
Пропуски в датасете geo_data_ 2 .

id         0
f0         0
f1         0
f2         0
product    0
dtype: int64
Статистика для датасета geo_data_ 2 .

                  f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.002023      -0.002081       2.495128      95.000000
std         1.732045       1.730417       3.473445      44.749921
min        -8.760004      -7.084020     -11.970335       0.000000
25%        -1.162288      -1.174820       

(None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None)

Как можно заметить все признаки, за исключением признака "id" являются численными (количественными). К тому же признак "id" не несет полезной для предсказания информации, поэтому удалим его из датасетов.  

In [5]:
geo_data_0 = geo_data_0.drop('id', axis = 1)
geo_data_1 = geo_data_1.drop('id', axis = 1)
geo_data_2 = geo_data_2.drop('id', axis = 1)

## Обучение и проверка модели

Разобьем данные на обучающую и валидационную выборки в соотношении 75:25.

In [6]:
features_0 = geo_data_0.drop("product", axis=1)
target_0 = geo_data_0["product"]
features_1 = geo_data_1.drop("product", axis=1)
target_1 = geo_data_1["product"]
features_2 = geo_data_2.drop("product", axis=1)
target_2 = geo_data_2["product"]

features_train_0, features_valid_0, target_train_0, target_valid_0 = train_test_split(
    features_0, target_0, test_size=0.25, random_state=12345
)
features_train_1, features_valid_1, target_train_1, target_valid_1 = train_test_split(
    features_1, target_1, test_size=0.25, random_state=12345
)
features_train_2, features_valid_2, target_train_2, target_valid_2 = train_test_split(
    features_2, target_2, test_size=0.25, random_state=12345
)

print("Размер features_train_0", features_train_0.shape)
print("Размер features_valid_0", features_valid_0.shape)

Размер features_train_0 (75000, 3)
Размер features_valid_0 (25000, 3)


Проведем масштабирвоание численых признаков.

In [7]:
pd.options.mode.chained_assignment = None
numeric = geo_data_0.columns.drop("product")

scaler = StandardScaler()
scaler.fit(features_train_0[numeric])

features_train_0[numeric] = scaler.transform(features_train_0[numeric])
features_valid_0[numeric] = scaler.transform(features_valid_0[numeric])

scaler = StandardScaler()
scaler.fit(features_train_1[numeric])

features_train_1[numeric] = scaler.transform(features_train_1[numeric])
features_valid_1[numeric] = scaler.transform(features_valid_1[numeric])

scaler = StandardScaler()
scaler.fit(features_train_2[numeric])

features_train_2[numeric] = scaler.transform(features_train_2[numeric])
features_valid_2[numeric] = scaler.transform(features_valid_2[numeric])

Напечатаем на экране средний запас предсказанного сырья и RMSE модели.

In [8]:
def pred(features_train, target_train, features_valid, target_valid):
    model = LinearRegression().fit(features_train, target_train)
    prediction_valid = model.predict(features_valid)
    return prediction_valid


predictions_valid_0 = pred(
    features_train_0, target_train_0, features_valid_0, target_valid_0
)
predictions_valid_1 = pred(
    features_train_1, target_train_1, features_valid_1, target_valid_1
)
predictions_valid_2 = pred(
    features_train_2, target_train_2, features_valid_2, target_valid_2
)

print(
    "Cредний запас предсказанного сырья источника № 0 = {:.2f}, RMSE = {:.3f}".format(
        predictions_valid_0.mean(),
        mean_squared_error(target_valid_0, predictions_valid_0, squared=False),
    )
)
print(
    "Cредний запас предсказанного сырья источника № 0 = {:.2f}, RMSE = {:.3f}".format(
        predictions_valid_1.mean(),
        mean_squared_error(target_valid_1, predictions_valid_1, squared=False),
    )
)
print(
    "Cредний запас предсказанного сырья источника № 0 = {:.2f}, RMSE = {:.3f}".format(
        predictions_valid_2.mean(),
        mean_squared_error(target_valid_2, predictions_valid_2, squared=False),
    )
)

Cредний запас предсказанного сырья источника № 0 = 92.59, RMSE = 37.579
Cредний запас предсказанного сырья источника № 0 = 68.73, RMSE = 0.893
Cредний запас предсказанного сырья источника № 0 = 94.97, RMSE = 40.030


Проанализируем результаты. Как мы видим, наибольшее средний запас предсказанного сырья имеет источники № 0 и 2, а также можно заметить, что RMSE у этих источников наименьший. Чем меньше значение RMSE тем меньше разница между спрогнозированными моделью  и фактическими значениями.

## Подготовка к расчёту прибыли

Все ключевые значения для расчётов сохраним в отдельных переменных

In [9]:
budget = 10e09 # бюджет на разработку скважин в регионе 
one_barrel = 450 # доход с одного бареля
income_from_product = 450 * 1e03 # доход с каждой единицы продукта
number_of_wells = 200 # количество скважин
n_points = 500
best_points = 200

Рассчитаем достаточный объём сырья для безубыточной разработки новой скважины. Сравним полученный объём сырья со средним запасом в каждом регионе. 

In [10]:
cost_one_well = budget / number_of_wells # определим стоимость одной скажины
volume_products = cost_one_well / income_from_product # определим рассчитаем достаточный объём сырья
print('Достаточный объём сырья = {:.0f}'.format(volume_products))

Достаточный объём сырья = 111


Сравнив достаточный объем сырья с средними запасами предсказанного сырья источника 0 и 2, 77,86 и 80,78, можно сделать вывод, что средних запасов предсказанного сырья источника имеют недостаточно для безубыточной разработки. Необходимо продолжить исследование для того, чтобы определить регионы, которые принесут наибольшую прибыль.

## Расчёт прибыли и рисков 

Напишем функцию для расчёта прибыли по выбранным скважинам и предсказаниям модели.Сперва выберем скважины с максимальными значениями предсказаний (200 скажин). Затем просуммируем целевое значение объёма сырья, соответствующее этим предсказаниям. После рассчитаем прибыль для полученного объёма сырья.

In [11]:
def profit(predictions, target):
    target_and_pred = (
        pd.DataFrame({"target": target, "preds": predictions})
        .sort_values(by="preds", ascending=False)
        .head(best_points)
    )
    target_sum = target_and_pred["target"].sum()
    total_income = target_sum * income_from_product - budget
    return total_income


print(
    "Прибыль c 200 лучших скажин с источника №:{:.0f}".format(
        profit(predictions_valid_0, target_valid_0)
    )
)
print(
    "Прибыль c 200 лучших скажин с источника №:{:.0f}".format(
        profit(predictions_valid_1, target_valid_1)
    )
)
print(
    "Прибыль c 200 лучших скажин с источника №:{:.0f}".format(
        profit(predictions_valid_2, target_valid_2)
    )
)

Прибыль c 200 лучших скажин с источника №:3320826043
Прибыль c 200 лучших скажин с источника №:2415086697
Прибыль c 200 лучших скажин с источника №:2710349964


Посчитаем риски и прибыль для каждого региона. Применим технику Bootstrap с 1000 выборок, чтобы найти распределение прибыли. Найдем среднюю прибыль, 95%-й доверительный интервал и риск убытков. 

In [12]:
state = np.random.RandomState(12345)


def bootstrap_profit(predictions, target, number):
    target_and_pred = pd.DataFrame(
        {"target": target, "preds": predictions}
    ).sort_values(by="preds", ascending=False)
    values = []
    for i in range(1000):
        target_and_pred_subsamples = (
            target_and_pred.sample(n=n_points, replace=True, random_state=state)
            .sort_values(by="preds", ascending=False)
            .reset_index()
            .head(best_points)
        )
        target_subsample = target_and_pred_subsamples["target"]
        predictions_subsample = target_and_pred_subsamples["preds"]
        target_sum = target_subsample.sum()
        total_income = target_sum * income_from_product - budget
        values.append(profit(predictions_subsample, target_subsample))
    values = pd.Series(values)
    risk = (values < 0).mean()
    lower = values.quantile(0.025)
    upper = values.quantile(0.975)
    mean = values.mean()
    return (
        print("Средняя прибыль источника №", number, ": {:.0f}".format(mean)),
        print(
            "95%-й доверительный интервал источника №",
            number,
            "от {:.0f} до {:.0f}".format(lower, upper),
        ),
        print("Убыток источника №", number, " {:.2%}".format(risk)),
        print(""),
    )

In [13]:
bootstrap_profit(predictions_valid_0, target_valid_0, 0)
bootstrap_profit(predictions_valid_1, target_valid_1, 1)
bootstrap_profit(predictions_valid_2, target_valid_2, 2)

Средняя прибыль источника № 0 : 381774792
95%-й доверительный интервал источника № 0 от -112683592 до 871816272
Убыток источника № 0  7.30%

Средняя прибыль источника № 1 : 438247743
95%-й доверительный интервал источника № 1 от 21812832 до 832341967
Убыток источника № 1  1.70%

Средняя прибыль источника № 2 : 382588613
95%-й доверительный интервал источника № 2 от -133252127 до 891051732
Убыток источника № 2  6.50%



(None, None, None, None)

Вывод: в данном проекте необходимо было определить регион, где добыча нефти принесёт наибольшую прибыль. Сперва были обработаны данные, а именно было определено отсутствие пропусков, удален ненужный для обучения модели и прогнозирования признак "id". Также для понимания наличия или отсутствия выбросов была создадана описательная статистика для всех датафреймов (для каждого источника или региона). В дальнейшем были проанализированы средний запас предсказанного сырья и RMSE модели для каждого источника и было обнаружено, что наименьший средний запас предсказанного сырья имеет источник №1. Остальные источники показали примерно равное значение среднего запаса. Также значение RMSE в источнике №1 было наибольшим, по сравнению с источниками № 0 и 2. Это говорит о том, что разница между предсказанными данными и фактическими является наибольшей среди всех участков. В дальнейшем было определен достаточный объём сырья = 111.  По каждому источнику были определны 200 самых лучших по продукту скважины и подсчитана прибыль. Получилось, что источник № 0 принесет с 200 самых продуктивных 3377406248 рублей, что является наилучшим показателем. источник № 2 имеет значение 2710295047, а № 0 2413656101. Рассчитав же такие параметры "средняя прибыль источника", "95%-й доверительный интервал" и "убыток источника" с помощью Bootstrap можно сделать вывод о том, что источник №1 имеет наилучшие показатели, так "средняя прибыль источника № 1" : 435891551, "95%-й доверительный интервал источника № 1" от 15337829 до 830892711, "убыток источника № 1"  1.80%. На основании полученных цифр рекомендую для добычи нефти источник №1.